### PR TITLE
Fix n+1 queries in frontend app - HomeController#index for promotions

### PR DESCRIPTION
**Pull request by Bharat311**
In the frontend app - on main homepage (HomeController#index) - there are a couple of promotion_rule and promotion queries being fired for each Spree::Product. The reason can be found [here](https://github.com/spree/spree/blob/master/core/app/helpers/spree/products_helper.rb#L62).

Fixed it by making the _possible_promotions_ a through relation on Spree::Product so that it can be  eager-loaded while loading products.

Similarly, fixed for ProductsController#index as well.

### CODE REVIEW

**Comment by priyank-gupta** [comment]
:+1: 
@damianlegawiec @Mafi88 ?

**Comment by Mafi88** [comment]
@Bharat311 Could you please address the 'align parameters' houndci issues?

@priyank-gupta, won't this new scope return duplicates?

**Comment by priyank-gupta** [comment]
@Mafi88 Not sure which scope your are addressing here but I think neither will return duplicates.

File: core/app/models/spree/product.rb [hunk 1]
```diff
@@ -36,6 +36,14 @@ class Product < Spree::Base
     has_many :product_promotion_rules, class_name: 'Spree::ProductPromotionRule'
     has_many :promotion_rules, through: :product_promotion_rules, class_name: 'Spree::PromotionRule'
 
+    has_many :promotions, through: :promotion_rules, class_name: 'Spree::Promotion'
```
Comment (Mafi88) [inline core/app/models/spree/product.rb]: @priyank-gupta I meant this association, shouldn't it include ```-> { uniq }```

Comment (priyank-gupta) [inline core/app/models/spree/product.rb]: @Mafi88 I think not, as has_many will return records for each entry in through association which will be unique already.

**Comment by Bharat311** [comment]
@Mafi88 - I'll be glad if you could please guide me about how i should align the parameters ?

I have tried to follow the same code styles used in the spree app but still the houndci raises a warning :disappointed: 

Thanks

File: core/app/models/spree/product.rb [hunk 2]
```diff
@@ -36,6 +36,14 @@ class Product < Spree::Base
...
+
+    has_many :possible_promotions,
+      -> { advertised.active },
```
Comment (priyank-gupta) [inline core/app/models/spree/product.rb]: @Bharat311 Try aligning method arguments by same indentation as first like `->  { advertised.active }` should align with :possible_promotions and respectively other arguments too. Hope this works.

**Comment by Mafi88** [comment]
@Bharat311 You can do it like this (which I think is easiest to read):

```
has_many :possible_promotions, -> { advertised.active }, through: :promotion_rules,
                                                         class_name: 'Spree::Promotion',
                                                         source: :promotion
```
or if you prefer splitting it into multiple lines, you can do:

```
has_many :possible_promotions,
         -> { advertised.active },
         through: :promotion_rules,
         class_name: 'Spree::Promotion',
         source: :promotion
```

like @priyank-gupta suggested.

Please rebase with latest master.

**Comment by Bharat311** [comment]
@Mafi88 @priyank-gupta - Updated the PR with said changes and rebased to latest master.

**Comment by Mafi88** [comment]
@Bharat311 Could you please address new houndci issues (similarly to the previous ones)?

**Comment by tanmay3011** [comment]
@priyank-gupta Addressed Hound Issues

**Comment by damianlegawiec** [comment]
@tanmay3011 @Bharat311 I can't seem to find in the new implementation logic for `reject(&:expired?)` - care to explain? :)

**Comment by tanmay3011** [comment]
@damianlegawiec  I see an extra `active` scope being used in possible_promotions association `-> { advertised.active }`. Which should work and cover old implementation of `reject(&:expired?)`

**Pull request merged**